corehq/apps/locations/bulk_management.py

```diff
@@ -0,0 +1,795 @@
+                )
+            allowed_to_codes = to_validator(lt.code)
+            if lt.expand_to and lt.expand_to not in allowed_to_codes:
+                errors.append(
+                    _(u"'{}' can't have '{}' as 'Expand To', valid options are '{}'")
+                    .format(lt.code, lt.expand_to, allowed_to_codes)
+                )
+
+        return errors
+
+    @memoized
+    def _validate_location_tree(self):
+        errors = []
+
+        def _validate_location(location):
+            loc_type = self.types_by_code.get(location.location_type)
+            if not loc_type:
+                return (_(
+                    u"Location '{}' in sheet points to a nonexistent or to be deleted location-type '{}'")
+                    .format(location.site_code, location.location_type))
+
+            parent = self.locations_by_code.get(location.parent_code)
+            if loc_type.parent_code == ROOT_LOCATION_TYPE:
+                if parent:
+                    return _(u"Location '{}' is a '{}' and should not have a parent").format(
+                        location.site_code, location.location_type)
+                else:
+                    return
+            else:
+                if not parent:
+                    return _(u"Location '{}' does not have a parent set or its parent is being deleted").format(
+                        location.site_code)
+            correct_parent_type = loc_type.parent_code
+            if parent == ROOT_LOCATION_TYPE or parent.location_type != correct_parent_type:
+                return _(u"Location '{}' is a '{}', so it should have a parent that is a '{}'").format(
+                    location.site_code, location.location_type, correct_parent_type)
+
+        for location in self.locations:
+            error = _validate_location(location)
+            if error:
+                errors.append(error)
+
+        return errors
+
+    @memoized
+    def _check_location_names(self):
+        locs_by_parent = defaultdict(list)
+        for loc in self.locations:
+            locs_by_parent[loc.parent_code].append(loc)
+        errors = []
+        for parent, siblings in locs_by_parent.items():
+            counts = Counter(l.name for l in siblings).items()
+            for name, count in counts:
+                if count > 1:
+                    errors.append(
+                        (_(u"There are {} locations with the name '{}' under the parent '{}'")
+                         .format(count, name, parent))
+                    )
+        return errors
+
+
+def new_locations_import(domain, excel_importer):
+    try:
+        importer = NewLocationImporter.from_excel_importer(domain, excel_importer)
+    except LocationExcelSheetError, e:
+        result = LocationUploadResult()
+        result.errors = [str(e)]
+        return result
+
+    result = importer.run()
+    return result
+
+
+def bulk_delete(objects):
+    # Given a list of existing SQL objects, bulk delete them
```

Comment:
These methods are interfaces to delete/update/create SQL objects in bulk. Currently they don't really do anything in bulk. There are some issues.
1. There isn't really a `bulk_delete` in Django.
2. Invoking `delete` on individual objects were propagating those changes to couch. Doing in bulk means, we have to address those propagations as well (this applies to other methods `bulk_create` and `bulk_update` as well)